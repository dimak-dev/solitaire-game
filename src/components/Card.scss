.card {
    width: 100%;
    height: 100%;
    display: inline-block;
    border-radius: 1rem;
    box-shadow: 1px 1px 1px 1px black;
    //border-style: solid;
    border-width: 1px;
    background: white;
    border-color: black;
    color: black;
    font-size: 3.5vw;

    &.red {
        color: red;
    }

    .flex-wrapper {
        display: flex;
        flex-direction: column;
        height: 100%;
        width: 100%;
    }

    .top-line {
        flex: 1;
    }

    .bottom-line {
        flex: 1;
        rotate: 180deg;
    }
}
