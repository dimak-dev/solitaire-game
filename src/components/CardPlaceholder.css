.card-placeholder {
    width: 14.2vw;
    max-width: 154px;
    height: 19.5vw;
    max-height: 209px;
    display: block;
    position: relative;
    border-radius: 1rem;
    background-color: rgba(25, 137, 234, 0.3);
    overflow: hidden;
}



.card-placeholder .icon-wrapper {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100%;
    height: 100%;
}

.card-placeholder .icon-wrapper .icon {
    color: rgba(255, 255, 255, 0.499);
    padding: 5px;
    display: inline-block;
    font-size: 3vw;
}

@media (min-width: 1080px) {
    .card-placeholder .icon-wrapper .icon {
        font-size: 32px;
    }
}