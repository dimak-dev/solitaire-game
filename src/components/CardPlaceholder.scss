.card-placeholder {
    width: 14.2vw;
    max-width: 154px;
    height: 19.5vw;
    max-height: 209px;
    display: block;
    position: relative;
    border-radius: 1rem;
    border-style: solid;
    border-color: rgba(25, 137, 234);
    border-width: 1px;
    box-sizing: border-box;
    background-color: rgba(25, 137, 234, 0.3);
    overflow: hidden;

    &.target {
        background-color: rgba(25, 137, 234, 0.5);
    }

    .icon-wrapper {
        display: flex;
        align-items: center;
        justify-content: center;
        width: 100%;
        height: 100%;

        .icon {
            color: rgba(255, 255, 255, 0.499);
            padding: 5px;
            display: inline-block;
            font-size: 3vw;

            @media (min-width: 1080px) {
                font-size: 32px;
            }
        }
    }
}
