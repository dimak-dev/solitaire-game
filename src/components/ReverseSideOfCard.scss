.reverse-side {
  height: 100%;
  width: 100%;
  background: #7f7f7f;
  border-radius: 1rem;
  border-color: black;
  border-style: solid;
  border-width: 1px;
  box-sizing: border-box;

}
