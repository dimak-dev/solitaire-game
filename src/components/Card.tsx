import React, {memo} from 'react';
import './Card.scss';
import {ECardSuit} from "Types/ECardSuit";
import {ECardValue} from "Types/ECardValue";
import cardValueToChar from "Utils/cardValueToChar";
import cardSuitToChar from "Utils/cardSuitToChar";
import {cardSuiteColor} from "Utils/cardSuiteColor";


interface ICardProps {
    suit: ECardSuit,
    value: ECardValue,
}

/**
 * Card component.
 *
 * Memoized for decreasing count of renders.
 */
// const Card = memo();

export default function Card({suit, value}: ICardProps) {
    const valueChar = cardValueToChar(value);
    const suitChar = cardSuitToChar(suit);
    const color = cardSuiteColor(suit);

    return (
        <div className={`card ${color}`}>
            <div className="flex-wrapper">
                <div className="top-line">
                    {valueChar}{suitChar}
                </div>
                <div className="bottom-line">
                    {valueChar}{suitChar}
                </div>
            </div>
        </div>
    )
};