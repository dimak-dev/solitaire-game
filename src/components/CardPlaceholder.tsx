import React from 'react';
import './CardPlaceholder.css';

interface ICardPlaceholderProps {
    children?: JSX.Element,
}

export default function CardPlaceholder({children}: ICardPlaceholderProps) {
    return (
        <div className="card-placeholder">

            {children || (
                <div className='icon-wrapper'>
                    <div className='icon'>♤ ♧<br/>♡ ♢</div>
                </div>
            )}
        </div>
    )
};