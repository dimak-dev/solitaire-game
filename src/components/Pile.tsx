import React from "react";
import {ICardInPile} from "Types/ICardInPile";
import {IPile} from "Types/IGameBoard";
import CardPlaceholder from "Components/CardPlaceholder";
import ReverseSideOfCard from "Components/ReverseSideOfCard";
import Card from "Components/Card";
import {ICard} from "Types/ICard";

interface IPileProps {
    cards: ICardInPile[]
    id: IPile['id']
    isTarget: boolean;
    onCardClick: (pileId: IPile['id'], card: ICard) => void;
    onTargetClick: (pileId: IPile['id']) => void;
}

export default function Pile({cards, id: pileId, isTarget, onCardClick, onTargetClick}: IPileProps) {
    const showEmptyPlaceholder = !cards.length;

    const hiddenCards = cards
        .filter(({isOpen}) => !isOpen)
        .map(({card}) => card);

    const openedCards = cards
        .filter(({isOpen}) => isOpen)
        .map(({card}) => card);


    return (
        <div className="pile" id={pileId}>
            {hiddenCards.map(({id: cardId}) =>
                <CardPlaceholder key={cardId}>
                    <ReverseSideOfCard/>
                </CardPlaceholder>
            )}

            {openedCards.map((card) =>
                <CardPlaceholder
                    key={card.id}
                    onClick={() => onCardClick(pileId, card)}
                >
                    <Card {...card}/>
                </CardPlaceholder>
            )}

            {(showEmptyPlaceholder || isTarget) && (
                <CardPlaceholder onClick={() => onTargetClick(pileId)}/>
            )}
        </div>
    )
}