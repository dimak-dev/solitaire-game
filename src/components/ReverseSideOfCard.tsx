import React from 'react';
import './ReverseSideOfCard.scss';

interface IReverseSideOfCardProps {
    onClick?: () => void;
}
export default function ReverseSideOfCard({onClick}: IReverseSideOfCardProps) {
    return (<div className="reverse-side" onClick={onClick}/>)
}