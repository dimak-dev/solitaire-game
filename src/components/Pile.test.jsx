import CardPlaceholder from "Components/CardPlaceholder";
import ReverseSideOfCard from "Components/ReverseSideOfCard";
import Card from "Components/Card";
import renderer from "react-test-renderer";
import Pile from "Components/Pile";
import {fireEvent, render} from "@testing-library/react";

jest.mock('Components/CardPlaceHolder');
jest.mock('Components/Card');
jest.mock('Components/ReverseSideOfCard');


describe('Pile component', () => {
    let testPile;
    let onCardClickMock;
    let onTargetClickMock;

    beforeAll(() => {
        CardPlaceholder.mockImplementation(
            ({children, onClick}) => (
                <div className={'card-placeholder-mock'} onClick={onClick}>{children}</div>
            )
        );

        Card.mockImplementation(
            ({id, suit, value}) => (
                <div className={'card-mock'} data-testid={id}>
                    id: {id}, suit: {suit}, value: {value}
                </div>
            )
        )

        ReverseSideOfCard.mockReturnValue(
             (
                <div
                    className={'reverse-side-card-mock'}
                    data-testid="reverse-side-card"
                >
                    reverse side
                </div>
            )
        );
    });

    beforeEach(() => {
        onCardClickMock = jest.fn();
        onTargetClickMock = jest.fn();
        testPile = {
            id: 'test-pile-id',
            cards: [],
            isTarget: false,
            onCardClick: onCardClickMock,
            onTargetClick: onTargetClickMock,
        };
    });

    test('Empty pile', () => {
        testPile.cards = [];
        const tree = renderer
            .create(<Pile {...testPile}/>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('One opened card on pile', () => {
        const card = {id: 'test-id', suit: 'test-suit', value: 'test-value'};
        testPile.cards = [{isOpen: true, card}];

        const tree = renderer
            .create(<Pile {...testPile}/>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('One hidden and one opened card on pile', () => {
        const card1 = {id: 'test-card-1', suit: 'suit-1', value: 'value-1'};
        const card2 = {id: 'test-card-2', suit: 'suit-2', value: 'value-2'};
        testPile.cards = [{isOpen: false, card: card1}, {isOpen: true, card: card2}];

        const tree = renderer
            .create(<Pile {...testPile}/>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('Pile is target', () => {
        const card1 = {id: 'test-card-1', suit: 'suit-1', value: 'value-1'};
        const card2 = {id: 'test-card-2', suit: 'suit-2', value: 'value-2'};
        testPile.cards = [{isOpen: false, card: card1}, {isOpen: true, card: card2}];
        testPile.isTarget = true;

        const tree = renderer
            .create(<Pile {...testPile}/>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    describe('onClick events', () => {
        let getByTestId;
        const card1 = {id: 'test-card-1', suit: 'suit-1', value: 'value-1'};
        const card2 = {id: 'test-card-2', suit: 'suit-2', value: 'value-2'};

        beforeEach(() => {
            testPile.cards = [{isOpen: false, card: card1}, {isOpen: true, card: card2}];
            getByTestId = render(<Pile {...testPile}/>).getByTestId;
        });

        test('click on hidden card don\'t fire onClick events', () => {
            fireEvent.click(getByTestId('reverse-side-card'));

            expect(onCardClickMock).not.toHaveBeenCalled();
            expect(onTargetClickMock).not.toHaveBeenCalled();
        });

        test('click on opened card fire only onCardClick', () => {
            fireEvent.click(getByTestId('test-card-2'));
            expect(onTargetClickMock).not.toHaveBeenCalled();

            expect(onCardClickMock).toHaveBeenCalledTimes(1);
            expect(onCardClickMock).toHaveBeenCalledWith('test-pile-id', card2);
        })
    });

});