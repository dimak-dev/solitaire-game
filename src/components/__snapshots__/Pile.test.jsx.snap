// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pile component Empty pile 1`] = `
<div
  className="pile"
  id="test-pile-id"
>
  <div
    className="card-placeholder-mock"
    onClick={[Function]}
  />
</div>
`;

exports[`Pile component One hidden and one opened card on pile 1`] = `
<div
  className="pile"
  id="test-pile-id"
>
  <div
    className="card-placeholder-mock"
  >
    <div
      className="reverse-side-card-mock"
      data-testid="reverse-side-card"
    >
      reverse side
    </div>
  </div>
  <div
    className="card-placeholder-mock"
    onClick={[Function]}
  >
    <div
      className="card-mock"
      data-testid="test-card-2"
    >
      id: 
      test-card-2
      , suit: 
      suit-2
      , value: 
      value-2
    </div>
  </div>
</div>
`;

exports[`Pile component One opened card on pile 1`] = `
<div
  className="pile"
  id="test-pile-id"
>
  <div
    className="card-placeholder-mock"
    onClick={[Function]}
  >
    <div
      className="card-mock"
      data-testid="test-id"
    >
      id: 
      test-id
      , suit: 
      test-suit
      , value: 
      test-value
    </div>
  </div>
</div>
`;

exports[`Pile component Pile is target 1`] = `
<div
  className="pile"
  id="test-pile-id"
>
  <div
    className="card-placeholder-mock"
  >
    <div
      className="reverse-side-card-mock"
      data-testid="reverse-side-card"
    >
      reverse side
    </div>
  </div>
  <div
    className="card-placeholder-mock"
    onClick={[Function]}
  >
    <div
      className="card-mock"
      data-testid="test-card-2"
    >
      id: 
      test-card-2
      , suit: 
      suit-2
      , value: 
      value-2
    </div>
  </div>
  <div
    className="card-placeholder-mock"
    onClick={[Function]}
  />
</div>
`;
